### PR TITLE
Fix matching of scenario outline with duplicate values with VsTest

### LINKED ISSUE #539: Test cases of Scenario Outlines with multiple (named) example sections not matched with results when using VsTest
Related to #507.

In a Scenario Outline with multiple example sets (both named and unnamed), where test case names are generated ending with `Variant`, the test results cannot be matched to the results in the VsTest result file. With the examples below, this issue can be reproduced.

**Named example sets:**
```Gherkin
Scenario Outline: Deal with multiple named example sections with tests generated as Variant
    Then the scenario will '<result>'

    Examples: First set
    | result |
    | pass   |
    | fail   | 
    | pass   |

    Examples: Second set
    | result       |
    | fail         |
    | inconclusive |
    | fail         | 
    | pass         |
```

**Unnamed example sets:**
```Gherkin
Scenario Outline: Deal with multiple example sections with tests generated ending in Variant
    Then the scenario will '<result>'

    Examples:
    | result |
    | pass   |
    | fail   | 
    | pass   |

    Examples:
    | result       |
    | fail         |
    | inconclusive |
    | fail         | 
    | pass         |
```

### LINKED ISSUE #543: VsTestScenarioOutlineExampleMatcher fails to match on examples with comma in the first column
Adding ".Replace(",", String.Empty)" to:

            var matchValue = exampleValues[0]
                .Replace(" ", string.Empty)
                .Replace(":", string.Empty)
                .Replace("\\", string.Empty)
                .Replace("(", string.Empty)
                .Replace(")", string.Empty)
                .Replace(".", "_")
                .Replace("!", string.Empty)
                .Replace("&", string.Empty)
                .ToUpperInvariant()
                .Replace("Ä", "A")
                .Replace("Ö", "O")
                .Replace("Ü", "U")
                .Replace("ß", "B")
                .Replace("æ", "ae")
                .Replace('ø', 'o')
                .Replace('å', 'a')
                .Replace("Æ", "AE")
                .Replace('Ø', 'O')
                .Replace('Å', 'A');

fixes the problem. Shall I create pull request?

### PR DESCRIPTION
**Pull request by rik-smeets**
Resolves #539 and resolves #543.

Shout-out to @jvandertil for doing part of this. @jvandertil, if you have some time, I'd appreciate if you'd have a look at it too!

## Provide scenarios in the test harness solution

- [x] Cucumber
- [x] CucumberJS
- [x] MsTest
- [x] NUnit (2.x)
- [x] NUnit3
- [x] SpecRun
- [x] xUnit (1.x)
- [x] xUnit2

## Add corresponding tests to the test framework unit tests in the main solution

- [x] CucumberJson from JS Results
- [x] CucumberJson from Ruby Results
- [x] MsTest
- [x] NUnit2
- [x] NUnit3
- [ ] SpecRun
- [x] VsTest
- [x] XUnit1
- [x] XUnit2

### CODE REVIEW

**Comment by dirkrombauts** [comment]
Cool ... this looks good - I like the code that is there.

I miss some code, though ... I like to make _really_ sure that all test result providers are in sync, so for that reason I would like to see the same test cases for all the other providers as well. In the best case that means just making the changes you did in `WhenParsingVsTestResultsFileWithIndividualResults.cs` to the corresponding files in the other providers. If we are lucky, the other providers handle this situation correctly. If not, more programming is needed. In any case this means extending the test harness.

Do you have the time and knowledge to do this? If yes, could you imagine doing this?

**Comment by rik-smeets** [comment]
Did a quick check and failing are:
- SpecRun
- XUnit1
- XUnit2

The other providers handle it correctly as is. Could be worse I guess ;-).

I'll take a look into it but I do have limited time. It doesn't help that I don't have experience with any of these test providers, so I'll first need to check how they generate the test cases and how the matching is performed now. I'll take a look at it this Wednesday and let you know how well it ends up.

If I conclude I can't fix it or it would take too much time for me, could we proceed with this PR in some other way? For example by creating issues for the aforementioned providers for these specific scenarios, so perhaps someone with more experience with that provider can fix it?

**Comment by dirkrombauts** [comment]
If you conclude you can't do all of it, we'll find some other way of continuing. The Cucumber and CucumberJS providers for example are always a pain - I won't hold it against you if you leave those alone ;-)

File: src/Pickles/Pickles.TestFrameworks/VsTest/VsTestScenarioOutlineExampleMatcher.cs [hunk 1]
```diff
@@ -18,30 +18,173 @@
...
-
 using PicklesDoc.Pickles.ObjectModel;
 
 namespace PicklesDoc.Pickles.TestFrameworks.VsTest
 {
     public class VsTestScenarioOutlineExampleMatcher : IScenarioOutlineExampleMatcher
     {
-        private static readonly Regex VariantRegex = new Regex(@"(.*)_Variant([\d*])", RegexOptions.Compiled | RegexOptions.IgnoreCase);
-        private const int VariantNumberGroup = 2;
+        private static readonly Regex VariantWithExampleGroupRegex = new Regex(@"(?:[\S]+)_ExampleSet([\d]+)_Variant([\d]+)$", RegexOptions.Compiled | RegexOptions.IgnoreCase);
```
Comment (jvandertil) [inline src/Pickles/Pickles.TestFrameworks/VsTest/VsTestScenarioOutlineExampleMatcher.cs]: This is different than the regex I initially created, I'll have a look tomorrow to see if this works for me as well.

Comment (jvandertil) [reply src/Pickles/Pickles.TestFrameworks/VsTest/VsTestScenarioOutlineExampleMatcher.cs]: It does give me some different behavior, I will check if I can come up with a test file that exposes these issues.

**Comment by jvandertil** [comment]
It seems that the test / example group names are generated differently from SpecFlow 1.9.0 (which is used in the test harness) and 2.4.0 which I am using.

**Comment by dirkrombauts** [comment]
While doing this, we could also add scenarios to cover cases like in #554

**Comment by dirkrombauts** [comment]
@rik-smeets I'm taking a go at continuing this pull request. So far, I've done some housekeeping (merging, updating libraries, creating a task list).

**Comment by dirkrombauts** [comment]
I finally had some time to investigate this again. It seems SpecRun does not seem to support this situation. Too bad.

**Comment by dirkrombauts** [comment]
@rik-smeets I know you've been waiting this for a long time. Thank you for your patience!

**Comment by dirkrombauts** [comment]
Released in version [v2.21.0](https://github.com/picklesdoc/pickles/releases/tag/v2.21.0).

**Pull request merged**